src/engine/renderer/tr_light.cpp

```diff
@@ -743,120 +743,159 @@ void R_SetupLightProjection( trRefLight_t *light )
-				Vector4Set( lightProject[ 1 ], up[ 0 ], up[ 1 ], up[ 2 ], 0 );
-				Vector4Set( lightProject[ 2 ], normal[ 0 ], normal[ 1 ], normal[ 2 ], 0 );
+					VectorScale( right, ( 0.5f * dist ) / rLen, right );
+					VectorScale( up, - ( 0.5f * dist ) / uLen, up );
+
+					PlaneSet( lightProject[ 0 ], right[ 0 ], right[ 1 ], right[ 2 ], 0 );
+					PlaneSet( lightProject[ 1 ], up[ 0 ], up[ 1 ], up[ 2 ], 0 );
+					PlaneSet( lightProject[ 2 ], normal[ 0 ], normal[ 1 ], normal[ 2 ], 0 );
+				}
 
 				// now offset to center
-				VectorCopy( light->l.projTarget, targetGlobal );
-				targetGlobal[ 3 ] = 1;
 				{
-					a = DotProduct4( targetGlobal, lightProject[ 0 ] );
-					b = DotProduct4( targetGlobal, lightProject[ 2 ] );
-					ofs = 0.5 - a / b;
+					vec3_t targetGlobal;
+					VectorSet( targetGlobal,
+						light->l.projTarget[ 0 ], light->l.projTarget[ 1 ],
+						light->l.projTarget[ 2 ] );
+
+					{
+						vec_t a = DotProduct( targetGlobal, lightProject[ 0 ].normal );
+						vec_t b = DotProduct( targetGlobal, lightProject[ 2 ].normal );
+						vec_t ofs = 0.5 - a / b;
 
-					Vector4MA( lightProject[ 0 ], ofs, lightProject[ 2 ], lightProject[ 0 ] );
+						VectorMA( lightProject[ 0 ].normal, ofs, lightProject[ 2 ].normal, lightProject[ 0 ].normal );
+					}
+
+					{
+						vec_t a = DotProduct( targetGlobal, lightProject[ 1 ].normal );
+						vec_t b = DotProduct( targetGlobal, lightProject[ 2 ].normal );
+						vec_t ofs = 0.5 - a / b;
+
+						VectorMA( lightProject[ 0 ].normal, ofs, lightProject[ 2 ].normal, lightProject[ 0 ].normal );
+					}
 				}
+
 				{
-					a = DotProduct4( targetGlobal, lightProject[ 1 ] );
-					b = DotProduct4( targetGlobal, lightProject[ 2 ] );
-					ofs = 0.5 - a / b;
+					vec3_t start;
+
+					if ( !VectorCompare( light->l.projStart, vec3_origin ) )
+					{
+						VectorCopy( light->l.projStart, start );
+					}
+					else
+					{
+						VectorClear( start );
+					}
+
+					vec3_t stop;
+
+					if ( !VectorCompare( light->l.projEnd, vec3_origin ) )
+					{
+						VectorCopy( light->l.projEnd, stop );
+					}
+					else
+					{
+						VectorCopy( light->l.projTarget, stop );
+					}
 
-					Vector4MA( lightProject[ 1 ], ofs, lightProject[ 2 ], lightProject[ 1 ] );
+					// Calculate the falloff vector
+					vec3_t falloff;
+
+					{
+						VectorSubtract( stop, start, falloff );
+
+						vec_t falloffLen = VectorNormalize( falloff );
+						light->falloffLength = falloffLen;
+
+						if ( falloffLen <= 0 )
+						{
+							falloffLen = 1;
+						}
+
+						//FIXME ?
+						VectorScale( falloff, 1.0f / falloffLen, falloff );
+					}
+
+					PlaneSet( lightProject[ 3 ], falloff, -DotProduct( start, falloff ) );
 				}
 
-				if ( !VectorCompare( light->l.projStart, vec3_origin ) )
+				// we want the planes of s=0, s=q, t=0, and t=q
+				plane_t *frustum = light->localFrustum;
+
 				{
-					VectorCopy( light->l.projStart, start );
+					frustum[ FRUSTUM_LEFT ] = lightProject[ 0 ];
```

Comment:
Remove random braces around single statements?

Reply:
Now done. I also found a couple of remaining “noisy comments” I deleted too.